contracts/tokens.js

```diff
@@ -105,6 +108,198 @@ actions.createSSC = async () => {
       }
     }
   }
+
+  /**
+   * Adds the table to store authorized contracts to do certain actions, i.e. transfer
+   */
+  tableExists = await api.db.tableExists('authorizations');
+  if (tableExists === false) {
+    await api.db.createTable('authorizations', ['account', 'contract', 'version', 'symbol', 'action', 'type']);
+  }
+};
+
+/**
+ * Adds an authorized delegatee via a contract to perform
+ * the specified type of action
+ */
+actions.addAuthorization = async (payload) => {
+  const {
+    contract,
+    version,
+    action,
+    symbol,
+    type,
+    callingContractInfo,
+  } = payload;
+  if (api.assert(!callingContractInfo, 'this action cannot be invoked by a contract')) {
+    if (api.assert(contract
+      && contract.length > 3
+      && contract.length < 50
+      && version
+      && typeof version === 'number'
+      && !api.BigNumber(version).isNaN()
+      && api.BigNumber(version).gte(1)
+      && action
+      && typeof action === 'string'
+      && type
+      && typeof type === 'string'
+      && AUTHORIZATION_TYPES.includes(type)
+      && symbol
+      && typeof symbol === 'string', 'invalid params')
+    ) {
+      const token = await api.db.findOne('tokens', { symbol });
+      if (api.assert(token !== null, 'symbol does not exist')) {
+        const authorization = await api.db.findOne('authorizations', {
+          account: api.sender,
+          contract,
+          version,
+          action,
+          symbol,
+        });
+        if (api.assert(authorization === null, 'authorization already exists')) {
+          await api.db.insert('authorizations', {
+            account: api.sender,
+            contract,
+            version,
+            action,
+            symbol,
+          });
+
+          api.emit('addAuthorization', {
+            account: api.sender, contract, version, symbol, action,
+          });
+
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+};
+
+/**
+ * Removes an authorized delegatee via a contract
+ */
+actions.removeAuthorization = async (payload) => {
+  const {
+    contract,
+    version,
+    action,
+    symbol,
+    type,
+  } = payload;
+  if (api.assert(contract
+    && contract.length > 3
+    && contract.length < 50
+    && version
+    && typeof version === 'number'
+    && !api.BigNumber(version).isNaN()
+    && api.BigNumber(version).gte(1)
+    && action
+    && typeof action === 'string'
+    && type
+    && typeof type === 'string'
+    && AUTHORIZATION_TYPES.includes(type)
+    && symbol
+    && typeof symbol === 'string', 'invalid params')
+  ) {
+    const token = await api.db.findOne('tokens', { symbol });
+    if (api.assert(token !== null, 'symbol does not exist')) {
+      const authorization = await api.db.findOne('authorizations', {
+        account: api.sender,
```

Comment:
when called from a contract, the "account" param should be a param that can be set by the contract and when called by a user it should be forced with api.sender. The reason for this is that if you call this action from another contract and the current user (api.sender) is not the account for which you want to remove the auth, this code won't remove any auth